### PR TITLE
encoder & signalling

### PR DESCRIPTION
**Pull request by robey**
make Encoder a trait so it can do streaming, and allow signals to be attached to outbound messages.

### CODE REVIEW

File: src/main/scala/com/twitter/naggati/Codec.scala [hunk 1]
```diff
@@ -42,30 +84,41 @@ object DontCareCounter extends (Int => Unit) {
...
 
   private var stage = firstStage
 
   private def buffer(context: ChannelHandlerContext) = {
-    ChannelBuffers.dynamicBuffer(context.getChannel().getConfig().getBufferFactory())
+    ChannelBuffers.dynamicBuffer(context.getChannel.getConfig.getBufferFactory)
   }
 
   // turn an Encodable message into a Buffer.
   override final def handleDownstream(context: ChannelHandlerContext, event: ChannelEvent) {
```
Comment (arya) [inline src/main/scala/com/twitter/naggati/Codec.scala:130]: I think this method can be replaced with writeRequested, and made simpler. Also, you can use super.writeRequested to continue sending downstream your new message.

File: src/main/scala/com/twitter/naggati/codec/MemcacheCodec.scala [hunk 2]
```diff
@@ -42,17 +46,35 @@ case class MemcacheResponse(line: String, data: Option[Array[Byte]]) {
...
-      (if (data.isDefined) (data.get.size + MemcacheCodec.END.size) else 0)
-    val buffer = ChannelBuffers.buffer(size)
-    buffer.writeBytes(lineData)
-    buffer.writeBytes(MemcacheCodec.CRLF)
-    data.foreach { x =>
-      buffer.writeBytes(x)
-      buffer.writeBytes(MemcacheCodec.END)
+  def writeAscii(nettyChannel: NettyChannel): Option[ChannelBuffer] = {
+    stream.foreach { channel =>
+      channel.respond { response =>
```
Comment (arya) [inline src/main/scala/com/twitter/naggati/codec/MemcacheCodec.scala]: I think some machinery around upgrading a channel to a streaming response, then downgrading it after the twitter-channel closes would be generally useful even if the current client/server won't exercise that hole.

Comment (robey) [inline src/main/scala/com/twitter/naggati/codec/MemcacheCodec.scala]: how would that work? keep state attached to the channel that will throw an error if anyone tries to write while the channel is active, and then change that state back to normal when the channel is closed?

File: src/main/scala/com/twitter/naggati/Codec.scala (hunk 1 above)
Comment (robey) [inline src/main/scala/com/twitter/naggati/Codec.scala:130]: i don't remember my original rationale, but it looks like FrameDecoder is a class, and so is SimpleChannelHandler, so i could extend one or the other but not both (lame).

**Pull request merged**